pkg/model/api/v1/datasource/http.go

```diff
@@ -0,0 +1,325 @@
+		return fmt.Errorf("basic_auth and bearer_token set at the same time")
+	}
+	return nil
+}
+
+type HTTPConfig struct {
+	// URL is the url required to contact the datasource
+	URL *url.URL `json:"url" yaml:"url"`
+	// The way the UI will contact the datasource. Or through the Backend or directly.
+	// By default, Access is set with the value 'server'
+	Access HTTPAccess `json:"access,omitempty" yaml:"access,omitempty"`
+	// AllowedEndpoints is a list of tuple of http method and http endpoint that will be accessible.
+	// These parameters are only used when access is set to 'server'
+	AllowedEndpoints []HTTPAllowedEndpoint `json:"allowed_endpoints,omitempty" yaml:"allowed_endpoints,omitempty"`
+	// Auth is holding any security configuration for the http configuration.
+	// When defined, it's impossible to set the value of Access with 'browser'
+	Auth *HTTPAuth `json:"auth,omitempty" yaml:"auth,omitempty"`
+	// Headers can be used to provide additional header that needs to be forwarded when requesting the datasource
+	// When defined, it's impossible to set the value of Access with 'browser'
+	Headers map[string]string `yaml:"headers,omitempty"`
+}
+
+// tmpHTTPConfig is only used to custom the json/yaml marshalling/unmarshalling step.
+// It shouldn't be used for other purpose.
+type tmpHTTPConfig struct {
+	URL              string                `json:"url" yaml:"url"`
+	Access           HTTPAccess            `json:"access,omitempty" yaml:"access,omitempty"`
+	AllowedEndpoints []HTTPAllowedEndpoint `json:"allowed_endpoints,omitempty" yaml:"allowed_endpoints,omitempty"`
+	Auth             *HTTPAuth             `json:"auth,omitempty" yaml:"auth,omitempty"`
+	Headers          map[string]string     `yaml:"headers,omitempty"`
+}
+
+func (h *HTTPConfig) MarshalJSON() ([]byte, error) {
+	urlAsString := ""
+	if h.URL != nil {
+		urlAsString = h.URL.String()
+	}
+	tmp := &tmpHTTPConfig{
+		URL:              urlAsString,
+		Access:           h.Access,
+		AllowedEndpoints: h.AllowedEndpoints,
+		Auth:             h.Auth,
+		Headers:          h.Headers,
+	}
+	return json.Marshal(tmp)
+}
+
+func (h *HTTPConfig) MarshalYAML() (interface{}, error) {
+	urlAsString := ""
+	if h.URL != nil {
+		urlAsString = h.URL.String()
+	}
+	tmp := &tmpHTTPConfig{
+		URL:              urlAsString,
+		Access:           h.Access,
+		AllowedEndpoints: h.AllowedEndpoints,
+		Auth:             h.Auth,
+		Headers:          h.Headers,
+	}
+	return tmp, nil
+}
+
+func (h *HTTPConfig) UnmarshalJSON(data []byte) error {
+	var tmp tmpHTTPConfig
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+	if err := h.validate(tmp); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *HTTPConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var tmp tmpHTTPConfig
+	if err := unmarshal(&tmp); err != nil {
+		return err
+	}
+	if err := h.validate(tmp); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *HTTPConfig) validate(conf tmpHTTPConfig) error {
```

Comment:
nit: Hm this method also seems to be setting variables on the `HTTPConfig` itself. Perhaps this would be better named `validateAndParse(...)` or something?

Reply:
mmm I don't know, this 'convention' is used on every object. Well it's fine to change it everywhere, perhaps I'm a bit too familiar that in a `validate` method you have code that is changing and setting data. It's a sort of way to validate the data, because at the end, well your data is valid and contained the proper value.
That's the way I see it. WDYT ?